interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
+	}
+
+	return nil
+}
+
+func validateWhereAttr(where string) error {
+	if !whereRegexp.MatchString(where) {
+		return fmt.Errorf(`mount-control "where" attribute is invalid`)
+	}
+
+	if !cleanSubPath(where) {
+		return fmt.Errorf(`mount-control "where" pattern is not clean: %q`, where)
+	}
+
+	if _, err := utils.NewPathPattern(where); err != nil {
+		return fmt.Errorf(`mount-control "where" attribute error: %v`, err)
+	}
+
+	return nil
+}
+
+func validateMountTypes(types []string) error {
+	for _, t := range types {
+		if !typeRegexp.MatchString(t) {
+			return fmt.Errorf(`mount-control filesystem type invalid: %q`, t)
+		}
+	}
+	return nil
+}
+
+func validateMountOptions(options []string) error {
+	for _, o := range options {
+		if !strutil.ListContains(allowedMountOptions, o) {
+			return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		}
+	}
+	return nil
+}
+
+func validateMountInfo(mountInfo *MountInfo) error {
+	if err := validateWhatAttr(mountInfo.what); err != nil {
+		return err
+	}
+
+	if err := validateWhereAttr(mountInfo.where); err != nil {
+		return err
+	}
+
+	if err := validateMountTypes(mountInfo.types); err != nil {
+		return err
+	}
+
+	if err := validateMountOptions(mountInfo.options); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *mountControlInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
+	if err := systemd.EnsureAtLeast(209); err != nil {
+		return err
+	}
+
+	if err := enumerateMounts(plug, validateMountInfo); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *mountControlInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	mountControlSnippet := bytes.NewBuffer(nil)
+	emit := func(f string, args ...interface{}) {
+		fmt.Fprintf(mountControlSnippet, f, args...)
+	}
+	snapInfo := plug.Snap()
+
+	numAddedEntries := 0
+	enumerateMounts(plug, func(mountInfo *MountInfo) error {
+		if numAddedEntries == 0 {
+			fmt.Fprintf(mountControlSnippet, `
+  # Rules added by the mount-control interface
+  capability sys_admin,  # for mount
+
+  owner @{PROC}/@{pid}/mounts r,
+  owner @{PROC}/@{pid}/mountinfo r,
+  owner @{PROC}/self/mountinfo r,
+
+  /{,usr/}bin/mount ixr,
+  /{,usr/}bin/umount ixr,
+  # mount/umount (via libmount) track some mount info in these files
+  /run/mount/utab* wrlk,
+`)
+		}
+
+		source := snapInfo.ExpandSnapVariables(mountInfo.what)
+		target := snapInfo.ExpandSnapVariables(mountInfo.where)
+
+		typeRule := ""
+		if len(mountInfo.types) > 0 {
+			typeRule = "fstype=(" + strings.Join(mountInfo.types, ",") + ")"
```

Comment:
some fstypes need to be disallowed such as at least `cgroup2` or `proc` since they could potentially be used for sandbox escape by mounting a new instance of those somewhere the snap has full arbitrary write access in $SNAP_COMMON

just looking at the list from /proc/filesystems, I think these should be disallowed:

```
lowerdir=... settings
ecryptfs
devpts
hugetlbfs
bpf
securityfs
tracefs
debugfs
cgroup2
cgroup
proc
sysfs
```

maybe actually it is easier to enumerate specifically what filesystems we support mounting like vfat, squashfs, ext*, zfs, tmpfs, etc.

also related question is fuse, should we allow mounting fuse here?

Reply:
I agree - it is easier to open up an allowlist than to close off a hole from a denylist being too permissive - can we please change the denylist to an allowlist? I realise this means we then perhaps also don't get support for new filesystems automatically as we have to expand the allowlist in each case but it does make it safer (plus it also means that if some new filesystem gets added to the kernel that itself would allow a sandbox escape, we haven't implicitly granted access to that either).

Reply:
I second going with an allow list